File: 141-linked-list-cycle/linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;

        if(head ==NULL || head->next == NULL){
            return false;
        }


        while(fast!=NULL){
            slow = slow->next;
            fast = fast->next;

            if(fast){
                fast=fast->next;
            }

            if(fast==slow){
                return true;
            }

        }

        if(fast!=NULL && fast==slow)
        {
            return true;
        }

        return false;
    }
};

File: 141-linked-list-cycle/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # has cycle or not ? 
        slow =head 
        fast = head

        while (fast and fast.next):
            slow = slow.next
            fast = fast.next
            if fast:
                fast = fast.next 
            
            if (fast==slow):
                return True

        
        return False